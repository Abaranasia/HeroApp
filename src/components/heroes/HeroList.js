import React, { useMemo } from 'react'
import { getHeroByPublisher } from '../../selectors/getHeroByPublisher';
import { HeroCard } from './HeroCard';

export const HeroList = ( {publisher} ) => {

    //const heroes = getHeroByPublisher ( publisher ); 
    /** Esta llamada es ok, pero sería más óptimo usarlo con useMemo para evitar llamar
     *  a esta función cada vez que accedemos aquí al regresar de una ficha de heroe */
    
     const heroes = useMemo(() => getHeroByPublisher ( publisher ), [ publisher ]);
    // Así solo ejecutaremos la función de nuevo cuando cambiemos de publisher

    return (
        <div className="row row-cols-1 row-cols-md-2 g-4 animate__animated animate__fadeIn">
            {
                heroes.map( hero => (
                    <HeroCard 
                        key={hero.id}
                        {...hero}
                    />    
                ))
            }
        </div>
    )
}
