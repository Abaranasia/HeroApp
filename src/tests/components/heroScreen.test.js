import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { HeroScreen } from '../../components/heroes/HeroScreen';

describe('Pruebas de heroScreen', () => {
/** Este componente requiere de props para funcionar, lo cual le llega desde heroList */
    const historyMock = {
        length: 10,
        goBack: jest.fn(),
        push: jest.fn(),
    }
    

    test('Debe mostrar el componente redirect si no hay argumentos en el URL', () => {

        const wrapper = mount ( // Debo usar mount porque es un HO Component
            <MemoryRouter initialEntries= {[ '/hero' ]}> 
                <HeroScreen history= { historyMock } />
            </MemoryRouter>
            ); //no recibe parámetros
            
        //expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true);
        //Por defecto, cargado sin parámetros, debería ir a redirect     
    });


    test('Debe mostrar la ficha de un heroe recibido por el URL', () => {
        
        const wrapper = mount ( // Debo usar mount porque es un HO Component
            <MemoryRouter initialEntries= {[ '/hero/marvel-spider' ]}> 
                <Route path="/hero/:heroId" component= {HeroScreen} />
            </MemoryRouter>
            ); //sí recibe parámetros

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.row').exists()).toBe(true) //prueba de que se dibuja ficha de héroe
    });  


    test('Debe probar el botón regresar (handleReturn) con push', () => {
        // forzamos el mock para que sea <=2 para que cumpla el if
        const historyMock = {
            length: 1,
            goBack: jest.fn(),
            push: jest.fn(),
        }

        const wrapper = mount ( 
            <MemoryRouter initialEntries= {[ '/hero/marvel-spider' ]}> 
                <Route 
                    path="/hero/:heroId" 
                    component= { () => <HeroScreen history= {historyMock}/> } 
                />
            </MemoryRouter>
            ); //sí recibe parámetros y necesitamos un history para el push
        
            wrapper.find('button').prop('onClick')();

            expect(historyMock.push).toHaveBeenCalledWith('/');
            expect(historyMock.goBack).not.toHaveBeenCalled();
    });


    test('Debe probar el botón regresar (handleReturn) con goBack', () => {
        // forzamos el mock para que sea >=2 para que cumpla el else

        // Aquí usaremos el historyMock declarado arriba con length: 10

        const wrapper = mount ( 
            <MemoryRouter initialEntries= {[ '/hero/marvel-spider' ]}> 
                <Route 
                    path="/hero/:heroId" 
                    component= { () => <HeroScreen history= {historyMock}/> } 
                />
            </MemoryRouter>
            ); //sí recibe parámetros y necesitamos un history para el push
        
            wrapper.find('button').prop('onClick')();

            expect(historyMock.goBack).toHaveBeenCalled();
            expect(historyMock.push).not.toHaveBeenCalled();
    });
});