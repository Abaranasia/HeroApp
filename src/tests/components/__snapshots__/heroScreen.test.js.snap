// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de heroScreen Debe mostrar la ficha de un heroe recibido por el URL 1`] = `
<div
  className="row mt-5 animate__animated animate__fadeInLeft"
>
  <div
    className="col-4"
  >
    <img
      alt="superhero"
      className="img-thumbnail"
      src="../assets/heroes/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-gropu list-group-flush"
    >
      <li
        className="list-group-item"
      >
         
        <b>
          Alter ego: 
        </b>
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
         
        <b>
          Publisher: 
        </b>
         
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First appearance: 
        </b>
        Amazing Fantasy #15
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
